src/lib/mina_net2/mina_net2.ml

```diff
@@ -1372,726 +121,419 @@ module Pubsub = struct
-        let%bind echo_handler =
-          handle_protocol b ~on_handler_error:`Raise ~protocol:"echo"
-            (fun stream ->
-              let r, w = Stream.pipes stream in
-              let%map () = Pipe.transfer r w ~f:Fn.id in
-              Pipe.close w ;
-              handler_finished := true)
-          |> Deferred.Or_error.ok_exn
-        in
-        let%bind stream =
-          open_stream c ~protocol:"echo" b_peerid >>| Or_error.ok_exn
-        in
-        let r, w = Stream.pipes stream in
-        Pipe.write_without_pushback w testmsg ;
-        Pipe.close w ;
-        (* HACK: let our messages send before we reset.
-           It would be more principled to add flushing to
-           the stream interface. *)
-        let%bind () = after (Time.Span.of_sec 1.) in
-        let%bind _ = Stream.reset stream in
-        let%bind msg = Pipe.read_all r in
-        (* give time for [a] to notice the reset finish. *)
-        let%bind () = after (Time.Span.of_sec 1.) in
-        let msg = Queue.to_list msg |> String.concat in
-        assert (String.equal msg testmsg) ;
-        assert !handler_finished ;
-        let%bind () = Protocol_handler.close echo_handler in
-        let%map () = shutdown () in
-        ()
+      match Hashtbl.find t.subscriptions subscription_id with
+      | Some (Subscription.E sub) ->
+          upon
+            (Subscription.handle_and_validate sub ~validation_expiration ~sender
+               ~data) (function
+            | `Validation_timeout ->
+                [%log' warn t.logger]
+                  "validation callback timed out before we could respond"
+            | `Decoding_error e ->
+                [%log' error t.logger]
+                  "failed to decode message published on subscription $topic \
+                   ($subscription_id): $error"
+                  ~metadata:
+                    [ ("topic", `String (Subscription.topic sub))
+                    ; ( "subscription_id"
+                      , `String (Subscription.Id.to_string subscription_id) )
+                    ; ("error", Error_json.error_to_yojson e)
+                    ] ;
+                Libp2p_helper.send_validation t.helper ~validation_id
+                  ~validation_result:ValidationResult.Reject
+            | `Validation_result validation_result ->
+                Libp2p_helper.send_validation t.helper ~validation_id
+                  ~validation_result)
+      | None ->
+          [%log' error t.logger]
+            "asked to validate message for unregistered subscription id \
+             $subscription_id"
+            ~metadata:
+              [ ( "subscription_id"
+                , `String (Subscription.Id.to_string subscription_id) )
+              ] )
+  (* A new inbound stream was opened *)
+  | IncomingStream m -> (
+      let open IncomingStream in
+      let stream_id = id_get m in
+      let protocol = protocol_get m in
+      let peer = Libp2p_ipc.unsafe_parse_peer (peer_get m) in
```

Comment:
Is unsafe necessary?

Reply:
The underlying `Network_peer.Peer.Id` module has not checks for format of peer ids, and this involves unsafely casting the string representation the libp2p_helper gives us into that type.

Reply:
It's transitively unsafe because this involves a call to `unsafe_parse_peer_id`